binder/src/main/java/io/grpc/binder/util/ServiceBinding.java

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.binder.util;
+
+import android.content.ComponentName;
+import android.content.Context;
+import android.content.Intent;
+import android.content.ServiceConnection;
+import android.os.IBinder;
+import androidx.annotation.AnyThread;
+import androidx.annotation.MainThread;
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.Status;
+import java.util.concurrent.Executor;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Manages an Android binding that's restricted to at most one connection to the remote Service.
+ *
+ * <p>A note on synchronization & locking in this class. Clients of this class are likely to manage
+ * their own internal state via synchronization. In order to avoid deadlocks, we must not hold any
+ * locks while calling observer callbacks.
+ *
+ * <p>For this reason, while internal consistency is handled with synchronization (the state field),
+ * consistency on our observer callbacks is ensured by doing everything on the application's main
```

Comment:
FYI, we tend to use SynchronizationContext for that purpose. But there could be other reasons (like onServiceConnected()) to use the app's main thread. Even if we like SynchronizationContext, it is probably best not to change it until more stuff is in-place.